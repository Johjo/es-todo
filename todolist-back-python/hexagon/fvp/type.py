from typing import NewType
from uuid import UUID

TaskKey = NewType('TaskKey', UUID)