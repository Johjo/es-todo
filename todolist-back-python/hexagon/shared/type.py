from datetime import date
from typing import NewType
from uuid import UUID

TaskKey = NewType('TaskKey', UUID)
TaskName = NewType('TaskName', str)
TaskOpen = NewType('TaskOpen', bool)
TaskExecutionDate = NewType('TaskExecutionDate', date)

TodolistName = NewType('TodolistName', str)
TodolistContext = NewType('TodolistContext', str)
TodolistContextCount = NewType('TodolistContextCount', int)