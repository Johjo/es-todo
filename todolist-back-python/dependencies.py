from dataclasses import dataclass, field, replace
from enum import Enum
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)



    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        assert (resource_type, resource) in self.factory, f"{resource_type.value} for {resource} must be injected first"
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

