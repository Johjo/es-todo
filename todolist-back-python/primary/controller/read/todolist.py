from abc import ABC, abstractmethod
from dataclasses import dataclass

from dependencies import Dependencies
from hexagon.shared.type import TaskKey, TodolistName, TodolistContext, TodolistContextCount, TaskOpen, TaskName


# todo : use a task key instead of an id
@dataclass
class Task:
    id: TaskKey
    name: TaskName
    is_open: TaskOpen

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name
        }


class TodolistSetReadPort(ABC):
    @abstractmethod
    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        pass

    @abstractmethod
    def all_by_name(self) -> list[TodolistName]:
        pass

    @abstractmethod
    def counts_by_context(self, todolist_name: TodolistName) -> list[tuple[TodolistContext, TodolistContextCount]]:
        pass

    @abstractmethod
    def all_tasks(self, todolist_name: TodolistName) -> list[Task]:
        pass


class TodolistReadController:
    def __init__(self, dependencies: Dependencies):
        self.dependencies = dependencies

    def all_todolist_by_name(self) -> list[str]:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.all_by_name()

    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.task_by(todolist_name=todolist_name, task_key=task_key)

    def counts_by_context(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.counts_by_context(todolist_name=todolist_name)

    def export_to_markdown(self):
        return ""

    def to_markdown(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return to_markdown(todolist_set.all_tasks(todolist_name))


def to_markdown(tasks: list[Task]) -> str:
    def task_to_markdown(task: Task) -> str:
        return f"- [{" " if task.is_open else "x"}] {task.name}"
    return "\n".join([task_to_markdown(task) for task in tasks])
