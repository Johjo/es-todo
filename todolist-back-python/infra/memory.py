from uuid import UUID

from expression import Option, Nothing, Some

from hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[str, TodolistSnapshot] = {}

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        if todolist_name not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[todolist_name])

    def save(self, todolist: TodolistSnapshot):
        self.all_todolist[todolist.name] = todolist

    def task_by(self, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[todolist_name].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self) -> list[str]:
        return [name for name in self.all_todolist.keys()]

    def all_tasks(self, todolist_name: str) -> list[TaskSnapshot]:
        if todolist_name not in self.all_todolist:
            return []
        return list(self.all_todolist[todolist_name].tasks)


