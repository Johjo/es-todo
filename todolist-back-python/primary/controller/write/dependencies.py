from abc import ABC
from collections import namedtuple
from dataclasses import dataclass, replace, field
from typing import Any

from hexagon.todolist.aggregate import TodolistSetPort
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.open_task import OpenTask


@dataclass(frozen=True)
class Dependencies(ABC):
    use_case_factory: dict[Any, Any] = field(default_factory=dict)
    adapter_factory: Any = None

    def get_adapter(self, port) -> Any:
        assert self.adapter_factory, f"adapter for {port} must be injected first"
        return self.adapter_factory(self)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return replace(self, use_case_factory={**self.use_case_factory, use_case: use_case_factory})


    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return replace(self, adapter_factory=adapter_factory)

    def get_use_case(self, use_case: Any) -> Any:
        assert use_case in self.use_case_factory, f"use_case for {use_case} must be injected first"
        return self.use_case_factory[use_case](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    factories = {
        TodolistCreate: todolist_create_factory,
        OpenTask : open_task_use_case_factory,
    }

    for use_case, factory in factories.items():
        dependencies = dependencies.feed_use_case(use_case, factory)
    return dependencies


def todolist_create_factory(dependencies: Dependencies):
    return TodolistCreate(dependencies.get_adapter(TodolistSetPort))

def open_task_use_case_factory(dependencies: Dependencies):
    return OpenTask(dependencies.get_adapter(TodolistSetPort))
