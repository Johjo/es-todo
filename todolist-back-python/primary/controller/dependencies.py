from functools import reduce
from typing import Any

from dependencies import Dependencies
from hexagon.fvp.read.which_task import WhichTaskQuery
from hexagon.fvp.write.cancel_priority import CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    use_cases_with_factory: list[Any] = [ResetFvpSession, CancelPriority, ChooseAndIgnoreTaskFvp, TodolistCreate,
                                         OpenTaskUseCase, CloseTask, RewordTask, ImportManyTask, WhichTaskQuery]

    def feed_use_case(dep: Dependencies, use_case: Any) -> Dependencies:
        return dep.feed_use_case(use_case=use_case, use_case_factory=use_case.factory)

    return reduce(feed_use_case, use_cases_with_factory, dependencies)


