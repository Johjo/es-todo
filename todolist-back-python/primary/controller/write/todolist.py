from dependencies import Dependencies
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.shared.type import TaskKey, TodolistName, TaskExecutionDate
from hexagon.fvp.write.cancel_priority import CancelPriority as Fvp_CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.postpone_task import PostPoneTask
from hexagon.todolist.write.reword_task import RewordTask
from test.secondary.todolist.test_external_todolist_markdown import MarkdownTodolist


class TodolistWriteController:
    def __init__(self, dependencies: Dependencies) -> None:
        self.dependencies = dependencies

    def create_todolist(self, todolist_name: str) -> None:
        use_case: TodolistCreate = self.dependencies.get_use_case(TodolistCreate)
        use_case.execute(todolist_name=todolist_name)

    def open_task(self, todolist_name: str, task_name: str):
        use_case: OpenTaskUseCase = self.dependencies.get_use_case(OpenTaskUseCase)
        use_case.execute(todolist_name=todolist_name, name=task_name)

    def close_task(self, todolist_name: str, task_key: TaskKey):
        use_case: CloseTask = self.dependencies.get_use_case(CloseTask)
        use_case.execute(todolist_name=todolist_name, key=task_key)

    def reword_task(self, todolist_name: str, task_key: TaskKey, new_name: str):
        use_case: RewordTask = self.dependencies.get_use_case(RewordTask)
        use_case.execute(todolist_name, task_key, new_name)

    def import_many_tasks_from_markdown(self, todolist_name: str, markdown: str):
        use_case: ImportManyTask = self.dependencies.get_use_case(ImportManyTask)
        external_todolist = MarkdownTodolist(markdown)
        use_case.execute(todolist_name, external_todolist)

    def choose_and_ignore_task(self, chosen_task: TaskKey, ignored_task: TaskKey):
        use_case: ChooseAndIgnoreTaskFvp = self.dependencies.get_use_case(ChooseAndIgnoreTaskFvp)
        use_case.execute(chosen_task_id=chosen_task, ignored_task_id=ignored_task)

    def cancel_priority(self, task_key: TaskKey):
        use_case: Fvp_CancelPriority = self.dependencies.get_use_case(Fvp_CancelPriority)
        use_case.execute(task_key)

    def reset_all_priorities(self) -> None:
        use_case: ResetFvpSession = self.dependencies.get_use_case(ResetFvpSession)
        use_case.execute()

    def postpone_task(self, name: TodolistName, key: TaskKey, execution_date: TaskExecutionDate):
        use_case: PostPoneTask = self.dependencies.get_use_case(PostPoneTask)
        use_case.execute(todolist_name=name, key=key, execution_date=execution_date)
