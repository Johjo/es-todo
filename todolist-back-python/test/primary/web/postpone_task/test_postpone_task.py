from datetime import datetime

from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter  # type: ignore
from expression import Some
from webtest import TestApp  # type: ignore

from dependencies import Dependencies
from hexagon.shared.type import TaskExecutionDate
from infra.memory import Memory
from primary.web.pages import bottle_config
from test.hexagon.todolist.fixture import TaskKeyGeneratorForTest
from test.fixture import TodolistFaker
from test.primary.web.fixture import CleanResponse


def test_postpone_task(memory: Memory, task_key_generator : TaskKeyGeneratorForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    initial_task = fake.a_task()
    today = datetime.today().date()
    expected_task = initial_task.having(execution_date=Some(TaskExecutionDate(today)))

    todolist = fake.a_todolist().having(tasks=(initial_task,))
    memory.save(todolist.to_snapshot())

    response = app.post(f'/todo/{todolist.name}/item/{expected_task.key}/postpone', {"execution_date": str(today)})
    assert CleanResponse(response).location() == f"/todo/{todolist.name}"
    assert response.status_code == 302
    assert memory.by(todolist.name).value == todolist.having(tasks=[expected_task]).to_snapshot()


def test_display_error_if_date_is_invalid(memory: Memory, task_key_generator : TaskKeyGeneratorForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    initial_task = fake.a_task(1).having(name="buy the milk")

    todolist = fake.a_todolist("todolist").having(tasks=[initial_task])

    memory.save(todolist.to_snapshot())

    response = app.post(f'/todo/{todolist.name}/item/{initial_task.key}/postpone', {"execution_date": ""})

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())



