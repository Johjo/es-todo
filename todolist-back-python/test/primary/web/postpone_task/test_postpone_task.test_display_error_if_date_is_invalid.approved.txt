b'<H2>  Reporter la t\xc3\xa2che </H2>

<p>buy the milk</p>



<form action="/todo/todolist/item/00000000-0000-0000-0000-000000000001/postpone?" method="post">

        <label for="execution_date">Date d&#039;execution</label>

        <input type="date" id="execution_date" name="execution_date"

               value="">

    <input type="submit" value="Reporter">

</form>

'
